### ISSUE TITLE
Rule properties missing in event hooks (Exception: TypeError: Cannot read property 'data' of undefined at Rule.get [as data]....)

### ISSUE THREAD

**Issue by salami-art** [mistic100/jQuery-QueryBuilder#829]
I enjoy the way this plug-in works, in fact I am trying to implement on top of it. From the documentation I understand that the room I have to implement custom features is mostly in event hooks, however, the parameters provided in these hooks are incomplete.

For example, I cannot access a rule's `data` or `error` from the `Rule` object provided in the `getRuleFilters` callback (or in any other as far as I know...). Instead, I only get the rule's reference with no properties initialized other than its `id` and `$el`.

```javascript
$('#builder').on('getRuleFilters.queryBuilder.filter', (filters, rule) => {
    // Shows all the properties of `Rule` after it's been initialized
    console.log(rule);
    // But actually, properties such as `data`, `error`, `filter`, `flags` are all undefined
    console.log(_.cloneDeep(rule));
});
```
<img width="768" alt="Schermata 2019-10-11 alle 12 22 56" src="https://user-images.githubusercontent.com/4851213/66644736-faf5cf00-ec21-11e9-9e9b-13c9ea4c2895.png">

TL;DR: How do I get the full `Rule` object in event hooks?

**Comment by mistic100** [comment]
It is totally normal that these properties are undefined or null in this event : it is called right after it's creation during the initialisation process.
I checked a few other events like "afterUpdateRuleOperator" and the object is correctly initialized.

As to why your debugger display errors for the Rule properties, I have no idea, I cannot reproduce on Chrome or Firefox.

**Comment by salami-art** [comment]
It's a deep-cloned version of the rule object. You say it's normal that these properties are not initialized but what if I initialize the rule with the api, for example:

```javascript
function makeNewRule() {
    let newGroup = $('#builder').queryBuilder('addGroup', ...);
    let newRule = $('#builder').queryBuilder('addRule', newGroup, { data: { filters: 'only_show_these_filters' } } );
}

// [....]

$('#builder').on('getRuleFilters.queryBuilder.filter', (filters, rule) => {
    rule.data; // `undefined`
});
$('#builder').on('afterUpdateRuleOperator.queryBuilder', (filters, rule) => {
    rule.data; // I have my { filters: 'only_show_these_filters' } now but the rule is already rendered...
});

```

I need the options I pass to the rule to be defined in subsequent callbacks, if I try to access `rule.data` in the `getRuleFilters` callback, I get `undefined`, but in other callbacks is there again. So where does the data go in between these two steps? How do I know if a rule carries certain data while populating filters ? Why should I wait until operators html is being generated?

**Comment by mistic100** [comment]
Could you debug from here https://github.com/mistic100/jQuery-QueryBuilder/blob/dev/src/core.js#L541 to know where the data is lost or not transmitted ?

I admit it should be there ('getRuleFilters' is triggered inside 'createRuleFilters' method)

**Comment by salami-art** [comment]
The problem I'm having is not only with the `data` object. 
I can bring up another example, this varies depending on the event I'm adding a handler on.
steps to reproduce:
- Initialize the plugin with one rule that has a `value` property with non null value
```javascript
$('#builder').queryBuilder({
//...
rules: {
    "condition": "AND",
    "rules": [
        {
            "id": "...",
            "field": "...",
            "type": "string",
            "input": "select",
            "operator": "equal",
            "value": "VALUE"
        }
    ],
    "valid": true
},
//...
})
``` 
- add a handler on the `afterCreateRuleInput` event
```javascript
$('#builder').on('afterCreateRuleInput.queryBuilder', function(e, rule) {
    console.log(rule.value) // => 'undefined'
});
```
This is annoying because it prevents me from restoring rules from a previous query when I use plug-ins.
